segmentation/include/pcl/segmentation/extract_clusters.h

```diff
@@ -41,10 +41,157 @@
 
 #include <pcl/pcl_base.h>
 
+#include <pcl/common/utils.h>
 #include <pcl/search/pcl_search.h>
 
 namespace pcl
 {
+
+namespace detail {
+template <typename PointT, typename Function>
+constexpr static bool is_functor_for_additional_filter_criteria_v =
+    pcl::is_invocable_r_v<bool,
+                          Function,
+                          const pcl::remove_cvref_t<pcl::PointCloud<PointT>>&,
+                          pcl::index_t,
+                          const pcl::remove_cvref_t<Indices>&,
+                          pcl::index_t>;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/** \brief Decompose a region of space into clusters based on the Euclidean distance between points
+  * \param it cloud iterator for iterating over the points
+  * \param cloud the point cloud message
+  * \param additional_filter_criteria a functor specifying a criterion that must be satisfied by all point added to the cluster
+  * \param tree the spatial locator (e.g., kd-tree) used for nearest neighbors searching
+  * \note the tree has to be created as a spatial locator on \a cloud and \a indices
+  * \param tolerance the spatial cluster tolerance as a measure in L2 Euclidean space
+  * \param clusters the resultant clusters containing point indices (as a vector of PointIndices)
+  * \param min_pts_per_cluster minimum number of points that a cluster may contain (default: 1)
+  * \param max_pts_per_cluster maximum number of points that a cluster may contain (default: max int)
+  * \warning The cloud/indices passed here must be the same ones used to build the tree.
+  *  For performance reasons PCL on warns if the sizes are not equal and no checks are performed to verify if the content is the same
+  * \ingroup segmentation
+  */
+  template <typename PointT, typename FunctorT> void
+   extractEuclideanClusters(
+      ConstCloudIterator<PointT> &it, const PointCloud<PointT> &cloud, FunctorT additional_filter_criteria,
```

Comment:
I'm not liking this pattern where you pass the cloud iterator as an argument, but still pass the whole point cloud as well to the function. Please change to the following.
```cpp
extractEuclideanClusters(const PointCloud<PointT> &cloud, const Indices &indices, [...]);
```

Overloads which don't have indices in their function signature, should supply empty indices to this method. You call the appropriate constructor for CloudIterator after checking if `indices` is empty or not.

Reply:
I think the `cloudIterator` is not well engineered enough for this.

Let's kill the idea of using the cloudIterator in public interface and hide it. It's still useful for removing common code, but beyond that, `cloud` is needed for size check as well as passing to the user's lambda

Reply:
> Overloads which don't have indices in their function signature, should supply empty indices to this method

That's a problematic behavior chain in pipelines. If after filtering/previous segmentation my indices became empty, now this clustering would run on the entire point cloud :scream: